File: 2021_finals/tasks/archaeologist/managers/stub.cpp
#include "archaeologist.h"
#include <bits/stdc++.h>
#include <unistd.h>
#include <fcntl.h>
using namespace std;

static constexpr char COMMAND_END = 0;
static constexpr char COMMAND_SETLIGHT = 1;
static constexpr char COMMAND_TAKEPATH = 2;

static ssize_t read_exact(int fd, char* ptr, size_t count) {
    size_t ret = count;
    while (count > 0) {
        const ssize_t r = read(fd, ptr, count);
        if (r <= 0) return r;
        ptr += r;
        count -= r;
    }
    return ret;
}

template <typename T>
static ssize_t read_exact_integral(int fd, T& arg) {
    return read_exact(fd, reinterpret_cast<char*>(&arg), sizeof(T));
}

static ssize_t read_exact_vector(int fd, vector<int>& vec) {
    int sz;
    int rsp = read_exact_integral(fd, sz);
    if (rsp <= 0) return rsp;
    vec.assign(sz, 0);
    if (sz == 0) return rsp;
    return rsp + read_exact(fd, reinterpret_cast<char*>(vec.data()), sizeof(int) * sz);
}

template <typename T>
static void write_integral(char*& buf, T val) {
    buf = copy_n(reinterpret_cast<char*>(&val), sizeof(val), buf);
}

static int input_fd, output_fd;

pair<int, vector<int>> take_path(int corridor) {
    const size_t buf_sz = 1 + sizeof(int);
    char* buffer = new char[buf_sz];
    char* buffer_end = buffer;
    *buffer_end++ = COMMAND_TAKEPATH;
    write_integral(buffer_end, corridor);
    assert(buffer + buf_sz == buffer_end);
    if (write(output_fd, buffer, buf_sz) != buf_sz) exit(0);
    int curr_room;
    vector<int> paths;
    if (read_exact_integral(input_fd, curr_room) <= 0) exit(0);
    if (read_exact_vector(input_fd, paths) <= 0) exit(0);
    return make_pair(curr_room, move(paths));
}

void set_light(int level) {
    const size_t buf_sz = 1 + sizeof(int);
    char* buffer = new char[buf_sz];
    char* buffer_end = buffer;
    *buffer_end++ = COMMAND_SETLIGHT;
    write_integral(buffer_end, level);
    assert(buffer + buf_sz == buffer_end);
    if (write(output_fd, buffer, buf_sz) != buf_sz) exit(0);
}

int main(int argc, char** argv) {
	fclose(stdin);
	fclose(stdout);
	fclose(stderr);
    assert(argc > 2);
    input_fd = open(argv[1], O_RDONLY);
    output_fd = open(argv[2], O_WRONLY);
    int arr_input[3];
    while (read_exact(input_fd, reinterpret_cast<char*>(arr_input), sizeof(arr_input)) > 0) {
        const int N = arr_input[0], K = arr_input[1], L = arr_input[2];
        vector<int> parent(N, 0);
        if (read_exact(input_fd, reinterpret_cast<char*>(parent.data()), sizeof(int) * N) <= 0) exit(0);
        int initial_level;
        if (read_exact_integral(input_fd, initial_level) <= 0) exit(0);
        vector<int> paths;
        if (read_exact_vector(input_fd, paths) <= 0) exit(0);
        archaeologist(N, K, L, move(parent), initial_level, move(paths));
        if (write(output_fd, &COMMAND_END, 1) != 1) exit(0);
    }
    close(output_fd);
    close(input_fd);
}

File: 2021_finals/tasks/archaeologist/managers/stub.java
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.nio.ByteOrder;

public class stub {
    private static final byte COMMAND_END = 0;
    private static final byte COMMAND_SETLIGHT = 1;
    private static final byte COMMAND_TAKEPATH = 2;
    
    private static BufferedInputStream reader;
    private static BufferedOutputStream writer;
    
    public static class RoomPathsPair {
        public int room;
        public int[] paths;
    }
    
    private static void readExact(BufferedInputStream reader, byte[] target) throws IOException {
        int count = target.length;
        int ret = count;
        int off = 0;
        while (count > 0) {
            int res = reader.read(target, off, count);
            if (res <= 0) throw new RuntimeException();
            count -= res;
            off += res;
        }
    }
    
    private static int readExactInt(BufferedInputStream reader) throws IOException {
        ByteBuffer buf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
        readExact(reader, buf.array());
        return buf.getInt();
    }
    
    private static int[] readExactVector(BufferedInputStream reader) throws IOException {
        int len = readExactInt(reader);
        if (len == 0) return new int[0];
        ByteBuffer buf2 = ByteBuffer.allocate(4*len).order(ByteOrder.LITTLE_ENDIAN);
        readExact(reader, buf2.array());
        IntBuffer ib = buf2.asIntBuffer();
        int[] arr = new int[ib.remaining()];
        ib.get(arr);
        return arr;
    }
    
    private static void writeInt(BufferedOutputStream writer, int val) throws IOException {
        ByteBuffer buf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
        buf.putInt(val);
        writer.write(buf.array());
    }
    
    public static RoomPathsPair take_path(int corridor) {
        try {
            writer.write(COMMAND_TAKEPATH);
            writeInt(writer, corridor);
            writer.flush();
            RoomPathsPair ret = new RoomPathsPair();
            ret.room = readExactInt(reader);
            ret.paths = readExactVector(reader);
            return ret;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
    
    public static void set_light(int level) {
        try {
            writer.write(COMMAND_SETLIGHT);
            writeInt(writer, level);
            writer.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
    
    public static void main(String[] args) throws IOException {
        System.in.close();
        System.out.close();
        System.err.close();
        reader = new BufferedInputStream(new FileInputStream(args[0]));
        writer = new BufferedOutputStream(new FileOutputStream(args[1]));
        while (true) {
            try {
                int N = readExactInt(reader);
                int K = readExactInt(reader);
                int L = readExactInt(reader);
                int[] parent = new int[N];
                ByteBuffer buf2 = ByteBuffer.allocate(4*N).order(ByteOrder.LITTLE_ENDIAN);
                readExact(reader, buf2.array());
                IntBuffer ib = buf2.asIntBuffer();
                ib.get(parent);
                int initial_level = readExactInt(reader);
                int[] paths = readExactVector(reader);
                (new Archaeologist()).archaeologist(N, K, L, parent, initial_level, paths);
                writer.write(COMMAND_END);
                writer.flush();
            } catch (RuntimeException e) {
                break;
            }
        }
    }
}
